Add warnings in is_lat_lon_match() in ph5tostationxml to fix issue

damhuonglan opened the pull request:
### What does this PR do?
Fix #363
Add check_lat_lon_eve() into ph5/utilities/validation.py to give warning for both ph5tostationxml and ph5validate. The following errors will be reported:
+ "Channel longitude seems to be 0. Is this correct???"
+ "Channel longitude 185.0 not in range [-180,180]"
+ "Channel longitude ABC is not a number."
+ "No Station location/X/units_s value found."
+ "No Channel latitude value found."
+ "Channel latitude 100.0 not in range [-90,90]"
+ "Channel latitude ABC is not a number."
+ "No Station location/Y/units_s value found."
+ "No Channel location/Z/value_d value found."
+ "No Station location/Z/units_s value found."

For ph5tostationxml, latitude and longitude will be checked for rect_intersection and radial_intersection with the following errors reported:
+ "Box intersection: Lat -107 not in range [34,40] or Lon 100.0 not in range [-111,-105]"
+ "lat,lon=35,182: radial intersection between a point radius boundary [0,3] and a lat/lon point [36,-107]"
These errors will not be reported in ph5_validate because it has no option to feed minlat, maxlat, minlon, maxlon, central lat/lon point, min/max distance from geographic point like in ph5tostationxml.

dsentinel commented:
This looks good.
Can you please add tests?
Does validate catch these and notify appropriately?

akram-pic commented:
I have tested "issue #363 #364" under “detailerr_ph5tostationxml” and here is the report:

On a sample master.ph5, I updated the array table to bogus lat/lon values listed below (only changes are listed):

station 102, channel 1, -190.0,degrees,97.0,degrees,
station 102, channel 2, -190.0,degrees,97.0,degrees
station 101, channel 1 ,-1019350529.0,degrees,336966937.0,degrees

So I changed Lat to 97 in station 102, and Lon to -190. and removed point in station 101 to create a very different value. 

and I get complaints from ph5tostationxml (listed at the end). number of error messages matches the wrong values, but error message only reads Lat out of bound, it says nothing about Longitude out of bound.

Error message is also confusing Array and station numbers. Should be station 101, array 001.

One more thing is that when you continue reading the ph5tostationxml output, you see that the reported values for the wrong channel lat/lon are not correctly reflected. So that is why I am confused of where "ph5tostationxml" is pulling the values from (look like probably form Z, and it populates it out). I have changed 102, ch[1,2] but channel Z is not changed. Same for 101, only channel 1 is changed in array table and ch2,Z are unchanged. 
 
Thank you for working on this.




(ph5) ph51@ph5-1:~/Desktop/ph51/Working/test-akram/detailerr_ph5tostationxml/test2$ ph5tostationxml -n master.ph5 -p ~/Desktop/ph51/Working/test-akram/detailerr_ph5tostationxml --level=channel -f TEXT
[2020-04-20 22:15:02,382] - ph5.clients.ph5tostationxml - ERROR: array 102, station 001, channel 1: Lat 97.0 not in range [-90,90]
[2020-04-20 22:15:02,382] - ph5.clients.ph5tostationxml - ERROR: array 102, station 001, channel 2: Lat 97.0 not in range [-90,90]
[2020-04-20 22:15:02,382] - ph5.clients.ph5tostationxml - ERROR: array 101, station 001, channel 1: Lat 336966937.0 not in range [-90,90]
[2020-04-20 22:15:02,401] - ph5.clients.ph5tostationxml - ERROR: array 102, station 001, channel 1: Lat 97.0 not in range [-90,90]
[2020-04-20 22:15:02,401] - ph5.clients.ph5tostationxml - ERROR: array 102, station 001, channel 2: Lat 97.0 not in range [-90,90]
[2020-04-20 22:15:02,401] - ph5.clients.ph5tostationxml - ERROR: array 101, station 001, channel 1: Lat 336966937.0 not in range [-90,90]
#Network|Station|Location|Channel|Latitude|Longitude|Elevation|Depth|Azimuth|Dip|SensorDescription|Scale|ScaleFreq|ScaleUnits|SampleRate|StartTime|EndTime
9L|102||DP1|33.696704|-101.935105|1000.0|0.0|0.0|0.0|Geo Space GS-30CT/FairfieldNodal ZLAND 3C|1029640000.0|40.0|M/S|500.0|2019-11-24T20:16:11|2019-11-25T21:39:40
9L|102||DP2|33.696704|-101.935105|1000.0|0.0|90.0|0.0|Geo Space GS-30CT/FairfieldNodal ZLAND 3C|1029640000.0|40.0|M/S|500.0|2019-11-24T20:16:11|2019-11-25T21:39:40
9L|102||DPZ|33.696704|-101.935105|1000.0|0.0|0.0|90.0|Geo Space GS-30CT/FairfieldNodal ZLAND 3C|1029640000.0|40.0|M/S|500.0|2019-11-24T20:16:11|2019-11-25T21:39:40
9L|101||DP2|33.696694|-101.935053|1000.0|0.0|90.0|0.0|Geo Space GS-30CT/FairfieldNodal ZLAND 3C|1029640000.0|40.0|M/S|500.0|2019-11-24T20:09:45|2019-11-25T21:39:53
9L|101||DPZ|33.696694|-101.935053|1000.0|0.0|0.0|90.0|Geo Space GS-30CT/FairfieldNodal ZLAND 3C|1029640000.0|40.0|M/S|500.0|2019-11-24T20:09:45|2019-11-25T21:39:53

damhuonglan commented:
@akram-pic 
> and I get complaints from ph5tostationxml (listed at the end). number of error messages matches the wrong values, but error message only reads Lat out of bound, it says nothing about Longitude out of bound.

Currently It only check one thing at at time which means it lat is wrong it will not check lon. It only check lon after you fix lat. But I can fix it. There will be 4 criterion to check: Lat in (-90,90), Lon in (-180,180), Box intersection, radial intersection. Then 4 errors might be logged for the same entry.

> Error message is also confusing Array and station numbers. Should be station 101, array 001.

Do you want the format to be: 
`station 001, array 101, channel 1: Lat 336966937.0 not in range [-90,90]?`
 or even no channel at all: `station 001, array 101: Lat 336966937.0 not in range [-90,90]`?

> One more thing is that when you continue reading the ph5tostationxml output, you see that the reported values for the wrong channel lat/lon are not correctly reflected. So that is why I am confused of where "ph5tostationxml" is pulling the values from (look like probably form Z, and it populates it out). I have changed 102, ch[1,2] but channel Z is not changed. Same for 101, only channel 1 is changed in array table and ch2,Z are unchanged.

In the report you posted, there were only errors list for:

+ Array 101, station 1, channel 1
+ Array 102, station1, channel 1,2

It seems they were repeated but nothing for channel 3 or channel Z. I don't know what do you mean by saying ph5tostationxml pulling values from channel Z. Please bold the part that makes you think that way and explain why.

damhuonglan commented:
@akram-pic 
> and I get complaints from ph5tostationxml (listed at the end). number of error messages matches the wrong values, but error message only reads Lat out of bound, it says nothing about Longitude out of bound.

Currently It only check one thing at at time which means it lat is wrong it will not check lon. It only check lon after you fix lat. But I can fix it. There will be 4 criterion to check: Lat in (-90,90), Lon in (-180,180), Box intersection, radial intersection. Then 4 errors might be logged for the same entry.

> Error message is also confusing Array and station numbers. Should be station 101, array 001.

Do you want the format to be: 
`station 001, array 101, channel 1: Lat 336966937.0 not in range [-90,90]?`
 or even no channel at all: `station 001, array 101: Lat 336966937.0 not in range [-90,90]`?

> One more thing is that when you continue reading the ph5tostationxml output, you see that the reported values for the wrong channel lat/lon are not correctly reflected. So that is why I am confused of where "ph5tostationxml" is pulling the values from (look like probably form Z, and it populates it out). I have changed 102, ch[1,2] but channel Z is not changed. Same for 101, only channel 1 is changed in array table and ch2,Z are unchanged.

In the report you posted, there were only errors list for:

+ Array 101, station 1, channel 1
+ Array 102, station1, channel 1,2

It seems they were repeated but nothing for channel 3 or channel Z. I don't know what do you mean by saying ph5tostationxml pulling values from channel Z. Please bold the part that makes you think that way and explain why.

damhuonglan commented:
@akram-pic : Thank you for reviewing this.
> and I get complaints from ph5tostationxml (listed at the end). number of error messages matches the wrong values, but error message only reads Lat out of bound, it says nothing about Longitude out of bound.

I modified the code so it will report all errors about lat, lon for an entry.

> Error message is also confusing Array and station numbers. Should be station 101, array 001.

I fixed this bug.

> One more thing is that when you continue reading the ph5tostationxml output, you see that the reported values for the wrong channel lat/lon are not correctly reflected. So that is why I am confused of where "ph5tostationxml" is pulling the values from (look like probably form Z, and it populates it out). I have changed 102, ch[1,2] but channel Z is not changed. Same for 101, only channel 1 is changed in array table and ch2,Z are unchanged.

When you compare carefully between array_t_001 and the result, you can see that it only skip the error entries, not that it pulls value from Z. Station 101 has two correct entries. Station 102 show 3 entries though it has two incorrect entries because there are totally 6 entries for 102, 4 corrects ones in which 2 are exactly the same. 

```
9L|102||DP1|33.696704|-101.935105|1000.0|
9L|102||DP2|33.696704|-101.935105|1000.0|
9L|102||DPZ|33.696704|-101.935105|1000.0|
9L|101||DP2|33.696694|-101.935053|1000.0|
9L|101||DPZ|33.696694|-101.935053|1000.0| 
```

So I will keep the way it is now because to show the error entries I have to change a lot of logic in the code which we are trying to avoid.

akram-pic commented:
I re-tested, here is observations:

1- I like the way it is outputting error in 'channel' and 'response' level, looks good (see error message below for reference). thank you!
2- Under 'station' and 'network'  level it only says one line now: "[2020-04-30 16:17:38,105] - ph5.clients.ph5tostationxml - ERROR: value 97.0 out of bounds (-90, 90)"

Is it possible to have it mention the station name? it would make life easier.

Thanks a lot for working on this Lan!


++++++++++++++++
ph5tostationxml -n master.ph5 -p ~/Desktop/ph51/Working/test-akram/detailerr_ph5tostationxml/test2 --level=channel -f TEXT
[2020-04-30 16:10:34,567] - ph5.clients.ph5tostationxml - ERROR: array 001, station 102, channel 1: Lat 97.0 not in range [-90,90]
[2020-04-30 16:10:34,567] - ph5.clients.ph5tostationxml - ERROR: array 001, station 102, channel 1: Lon -190.0 not in range [-180,180]
[2020-04-30 16:10:34,567] - ph5.clients.ph5tostationxml - ERROR: array 001, station 102, channel 2: Lat 97.0 not in range [-90,90]
[2020-04-30 16:10:34,567] - ph5.clients.ph5tostationxml - ERROR: array 001, station 102, channel 2: Lon -190.0 not in range [-180,180]
[2020-04-30 16:10:34,567] - ph5.clients.ph5tostationxml - ERROR: array 001, station 101, channel 1: Lat 336966937.0 not in range [-90,90]
[2020-04-30 16:10:34,567] - ph5.clients.ph5tostationxml - ERROR: array 001, station 101, channel 1: Lon -1019350529.0 not in range [-180,180]
Check logs for errors. Please keep in mind that the entries with errors will not be displayed in the result.
Hit Enter/Return to continue.
#Network|Station|Location|Channel|Latitude|Longitude|Elevation|Depth|Azimuth|Dip|SensorDescription|Scale|ScaleFreq|ScaleUnits|SampleRate|StartTime|EndTime
9L|102||DP1|33.696704|-101.935105|1000.0|0.0|0.0|0.0|Geo Space GS-30CT/FairfieldNodal ZLAND 3C|1029640000.0|40.0|M/S|500.0|2019-11-24T20:16:11|2019-11-25T21:39:40

damhuonglan commented:
> 2- Under 'station' and 'network' level it only says one line now: "[2020-04-30 16:17:38,105] - ph5.clients.ph5tostationxml - ERROR: value 97.0 out of bounds (-90, 90)"

Thanks for the review, @akram-pic . It's a bug in the code. I fixed that. Please review this again. Thank you.

akram-pic commented:
I checked the latest changes to the code at network, station, channel, and response level and they all output the correct error (identical error message at all levels, which is good). 

Thank you.



ph5tostationxml -n master.ph5 -p ~/Desktop/ph51/Working/test-akram/detailerr_ph5tostationxml/test2 --level=station -f TEXT
[2020-05-11 17:08:41,440] - ph5.clients.ph5tostationxml - ERROR: array 001, station 102, channel 1: Lat 97.0 not in range [-90,90]
[2020-05-11 17:08:41,440] - ph5.clients.ph5tostationxml - ERROR: array 001, station 102, channel 1: Lon -190.0 not in range [-180,180]
[2020-05-11 17:08:41,440] - ph5.clients.ph5tostationxml - ERROR: array 001, station 102, channel 2: Lat 97.0 not in range [-90,90]
[2020-05-11 17:08:41,440] - ph5.clients.ph5tostationxml - ERROR: array 001, station 102, channel 2: Lon -190.0 not in range [-180,180]
[2020-05-11 17:08:41,440] - ph5.clients.ph5tostationxml - ERROR: array 001, station 101, channel 1: Lat 336966937.0 not in range [-90,90]
[2020-05-11 17:08:41,440] - ph5.clients.ph5tostationxml - ERROR: array 001, station 101, channel 1: Lon -1019350529.0 not in range [-180,180]
Check logs for errors. Please keep in mind that the entries with errors will not be displayed in the result.

dsentinel commented on ph5/clients/ph5tostationxml.py:
```diff
@@ -524,7 +519,8 @@ def get_response_inv(self, obs_channel):
     def create_obs_network(self):
         obs_stations = self.read_stations()
-        for errmsg, logtype in self.unique_errors:
+        unique_errors = sorted(list(self.unique_errors))
+        for errmsg, logtype in unique_errors:
```
```suggestion
        for errmsg, logtype in sorted(list(self.unique_errors)):
```

dsentinel commented on ph5/clients/tests/test_ph5tostationxml.py:
```diff
@@ -223,8 +223,9 @@ def test_create_obs_network(self):
             log.setLevel(logging.WARNING)
             self.parser.create_obs_network()
+            errmsgs = sorted(self.errmsgs)
             for i in range(len(self.errmsgs)):
```
```suggestion
```

damhuonglan replied:
I don't understand. Why do you remove the for loop? Do you mean change lines from 226 to 228 to: 
```
            for i in range(len(self.errmsgs)):
                self.assertEqual(log.records[i].msg, sorted(errmsgs)[i])
```

dsentinel commented on ph5/clients/tests/test_ph5tostationxml.py:
```diff
@@ -223,8 +223,9 @@ def test_create_obs_network(self):
+            errmsgs = sorted(self.errmsgs)
             for i in range(len(self.errmsgs)):
-                self.assertEqual(log.records[i].msg, self.errmsgs[i])
+                self.assertEqual(log.records[i].msg, errmsgs[i])
```
```suggestion
      self.assertEqual(set(rec.msg for rec in log.records), set(self.errmsgs))
```

damhuonglan replied:
Thank you for your suggestion. I have never tried to use set before.

akram-pic commented:
I have tested the followings: ph5_validate against a faulty master file that few stations have lat/lon out of bound.

**ph51@ph5-1:~/Desktop/ph51/Working/test-akram/detailerr_ph5tostationxml/test2$ ph5_validate -n master-test-branch.ph5 -p ~/Desktop/ph51/completed/20-009_Playa2** 

See results below, 

ph5_validate was able to catch and report for the faulty station/channel. Although I think it will be more appropriate to have it as "ERROR" instead of  "WARNING". 

Thank you  

(ph5) ph51@ph5-1:~/Desktop/ph51/Working/test-akram/detailerr_ph5tostationxml/test2$ more ph5_validate.log
-=-=-=-=-=-=-=-=-
Station 102 Channel 1
0 error, 4 warning, 0 info
-=-=-=-=-=-=-=-=-
WARNING: Channel longitude -190.0 not in range [-180,180]
WARNING: Channel latitude 97.0 not in range [-90,90]
WARNING: Station 102 [1574626571, 1574717980] is repeated 1 time(s)
WARNING: Data exists after pickup time: 758 seconds.
-=-=-=-=-=-=-=-=-
Station 102 Channel 2
0 error, 4 warning, 0 info
-=-=-=-=-=-=-=-=-
WARNING: Channel longitude -190.0 not in range [-180,180]
WARNING: Channel latitude 97.0 not in range [-90,90]
WARNING: Station 102 [1574626571, 1574717980] is repeated 1 time(s)
WARNING: Data exists after pickup time: 758 seconds.
-=-=-=-=-=-=-=-=-
Station 102 Channel 3
0 error, 2 warning, 0 info
-=-=-=-=-=-=-=-=-
WARNING: Station 102 [1574626571, 1574717980] is repeated 1 time(s)
WARNING: Data exists after pickup time: 758 seconds.
-=-=-=-=-=-=-=-=-
Station 101 Channel 1
0 error, 3 warning, 0 info
-=-=-=-=-=-=-=-=-
WARNING: Channel longitude -1019350529.0 not in range [-180,180]
WARNING: Channel latitude 336966937.0 not in range [-90,90]
WARNING: Data exists after pickup time: 375 seconds.
-=-=-=-=-=-=-=-=-
Station 101 Channel 2
0 error, 1 warning, 0 info
-=-=-=-=-=-=-=-=-
WARNING: Data exists after pickup time: 375 seconds.
-=-=-=-=-=-=-=-=-
Station 101 Channel 3
0 error, 1 warning, 0 info
-=-=-=-=-=-=-=-=-
WARNING: Data exists after pickup time: 375 seconds.
-=-=-=-=-=-=-=-=-

damhuonglan commented:
> WARNING: Channel longitude -190.0 not in range [-180,180]
> WARNING: Channel latitude 97.0 not in range [-90,90]
> WARNING: Station 102 [1574626571, 1574717980] is repeated 1 time(s)
> WARNING: Data exists after pickup time: 758 seconds.

Do you want all of these to be errors or which ones?

dsentinel commented:
@damhuonglan , please revert "hit  enter to continue". I think this will certainly break the DMC's workflow
